### PR TITLE
bpo-35942: Improve the error message if __fspath__ returns invalid types in path_converter

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue35942

### CODE REVIEW

**Comment by ZackerySpytz** [comment]
The issue number is wrong.

File: Modules/posixmodule.c [hunk 1]
```diff
@@ -954,20 +954,15 @@ path_converter(PyObject *o, void *p)
...
-        if (NULL == func) {
-            goto error_format;
-        }
-        /* still owns a reference to the original object */
-        Py_DECREF(o);
-        o = _PyObject_CallNoArg(func);
-        Py_DECREF(func);
-        if (NULL == o) {
+        res = PyOS_FSPath(o);
+        if (res == NULL){
```
Comment (ZackerySpytz) [inline Modules/posixmodule.c]: This doesn't comply with PEP 7.

Comment (pablogsal) [reply Modules/posixmodule.c]: Fixed!

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
I don't like this. It makes error messages for incorrect type of arguments worse.

Change only error messages for `__fspath__()` returning incorrect result.

**Comment by pablogsal** [comment]
@serhiy-storchaka I have changed the approach to only correct messages for `__fspath__()` returning an incorrect result.

I have made the requested changes; please review again

File: Modules/posixmodule.c [hunk 2]
```diff
@@ -955,19 +955,34 @@ path_converter(PyObject *o, void *p)
...
+            o = res;
             goto error_exit;
         }
-        else if (PyUnicode_Check(o)) {
+
+        /* still owns a reference to the original object */
+        Py_DECREF(o);
+        o = res;
+
+        if (PyUnicode_Check(o)) {
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c]: Move this up, before line 971 (using `res` instead of `o`).

File: Modules/posixmodule.c [hunk 3]
```diff
@@ -955,19 +955,34 @@ path_converter(PyObject *o, void *p)
...
             goto error_format;
         }
-        /* still owns a reference to the original object */
-        Py_DECREF(o);
-        o = _PyObject_CallNoArg(func);
+        res = _PyObject_CallNoArg(func);
         Py_DECREF(func);
-        if (NULL == o) {
+        if (NULL == res) {
+            Py_DECREF(o);
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c]: This line and the next one are not needed.

File: Modules/posixmodule.c [hunk 4]
```diff
@@ -955,19 +955,34 @@ path_converter(PyObject *o, void *p)
...
+            o = res;
+            goto error_exit;
+        }
+        if (!(PyUnicode_Check(res) || PyBytes_Check(res))) {
+            PyErr_Format(PyExc_TypeError,
+                 "expected %.200s.__fspath__() to return str or bytes, "
+                 "not %.200s", Py_TYPE(o)->tp_name,
+                 Py_TYPE(res)->tp_name);
+            Py_DECREF(o);
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c]: Replace this line and the next one with just `Py_DECREF(res)`.

File: Modules/posixmodule.c [hunk 5]
```diff
@@ -955,19 +955,34 @@ path_converter(PyObject *o, void *p)
         /* Inline PyOS_FSPath() for better error messages. */
         _Py_IDENTIFIER(__fspath__);
         PyObject *func = NULL;
+        PyObject *res = NULL;
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c]: No need to initialize `res` and `func` with NULL.

**Comment by pablogsal** [comment]
@serhiy-storchaka Thank you very much for the thorough review! I have addressed your feedback in 477daf8.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
I think this bugfix is forth a news entry.

**Review by ambv — APPROVED** [review]
Thanks for jumping on this so quickly, @pablogsal!

File: Modules/posixmodule.c [hunk 6]
```diff
@@ -954,28 +954,35 @@ path_converter(PyObject *o, void *p)
...
+            PyErr_Format(PyExc_TypeError,
+                 "expected %.200s.__fspath__() to return str or bytes, "
+                 "not %.200s", Py_TYPE(o)->tp_name,
+                 Py_TYPE(res)->tp_name);
+            Py_DECREF(res);
+            goto error_exit;
         }
+
+        /* still owns a reference to the original object */
+        Py_DECREF(o);
```
Comment (eamanu) [inline Modules/posixmodule.c:984]: A little question, Why don't make `Py_DECREF(res);` too?

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**